src/__flow-tests__/generics-tests.js

```diff
@@ -0,0 +1,272 @@
+
+  // We'll need the type we cast from to also be a more interesting type than
+  // `empty`, as a value of type `empty` causes Flow to just declare victory:
+  (magic: IsSupertype<number, mixed>);
+  // This is OK in practice because in real life there are no values of type
+  // `empty`; if you have an expression of type `empty`, you're already
+  // asserting falsehood.  (And it's perfectly OK from a theoretical
+  // perspective for basically the same reason said differently: `empty` is
+  // the type interpretation of logical falsity, and ex falso quodlibet.)
+
+  // Test the basics, relative to a primitive type.
+  (1: IsSupertype<number, empty>);
+  (1: IsSupertype<number, number>);
+  // $FlowExpectedError[incompatible-type-arg]
+  (1: IsSupertype<number, mixed>);
+  // $FlowExpectedError[incompatible-type-arg]
+  (1: IsSupertype<number, string>);
+
+  // Test the resulting type (as opposed to whether the type is instantiable
+  // at all, which is the main point of `IsSupertype`.)
+  (magic: IsSupertype<number, number>);
+  // $FlowExpectedError[incompatible-cast]
+  ((1: mixed): IsSupertype<number, number>);
+  // $FlowExpectedError[incompatible-cast]
+  ('': IsSupertype<number, number>);
+
+  // `empty` isn't a supertype of anything but itself.  It's a bit hard to
+  // test this fact directly, because of the above-mentioned issue that
+  // casting a value of type `empty` causes Flow to just declare victory:
+  (magic: IsSupertype<empty, number>);
+  // but, again, in reality there are no values of type `empty` anyway.
+  // If we try casting from another type, we get an error as you'd hope:
+  // $FlowExpectedError[incompatible-cast]
+  // $FlowExpectedError[incompatible-type-arg]
+  (1: IsSupertype<empty, number>);
+  // and in fact two of them-- one for the cast because nothing can be cast
+  // to `empty`, and one for instantiating IsSupertype in the first place.
+  // prettier-ignore
+  ( // $FlowExpectedError[incompatible-cast]
+    1:
+    // $FlowExpectedError[incompatible-type-arg]
+    IsSupertype<empty, number>);
+
+  // Test on unions.
+  (1: IsSupertype<number | void, number>);
+  // $FlowExpectedError[incompatible-type-arg]
+  (1: IsSupertype<number, number | void>);
+}
+
+function test_IsSupertype_object_types() {
+  // A value to cast.
+  const a: {| a: number |} = { a: 1 };
+  // Check that the casts don't themselves cause errors,
+  // so the errors we find below really do come from IsSupertype.
+  (a: { a: number, ... });
+  (a: {| a: number |});
+  (a: {| +a: number |});
+
+  // Exact <: inexact, strictly.
+  (a: IsSupertype<{ a: number, ... }, { a: number, ... }>);
+  (a: IsSupertype<{| a: number |}, {| a: number |}>);
+  (a: IsSupertype<{ a: number, ... }, {| a: number |}>);
+  // $FlowExpectedError[incompatible-exact]
+  (a: IsSupertype<{| a: number |}, { a: number, ... }>);
+
+  // Read-only properties are covariant.
+  (a: IsSupertype<{| +a: number |}, {| +a: empty |}>);
```

Comment:
Could put, e.g., `{| +a: 3 |}` instead of `{| +a: empty |}`, since we don't mean to invoke all the reasons why `empty` is special here.

Reply:
Hmm, well, this line is meant to be read together with the one below it (which make up a little stanza):
```js
  // Read-only properties are covariant.
  (a: IsSupertype<{| +a: number |}, {| +a: empty |}>);
  // $FlowExpectedError[incompatible-type-arg]
  (a: IsSupertype<{| +a: number |}, {| +a: mixed |}>);
```

The types `empty` and `mixed` are equally special, and in inverse ways: `empty` is a subtype of everything (can flow to anything), and `mixed` is a supertype of everything (anything can flow to it.)

So it's true that `empty` is special, but that specialness is really just a pure form of what this test needs here: it wants some type T that's clearly a subtype of `number`, in order to test that that implies `{| +a: T |}` is a subtype of `{| +a: number |}`. (And more generally: if `S` supertype `T` then `{| +a: S |}` supertype `{| +a: T |}`.)

We saw in `test_IsSupertype` above that if you have `empty` appearing at the input end of a data flow, then Flow may stop caring what type is at the output end, and not notice even if that latter type has a type expression like `IsSupertype<empty, number>` that should be an error. But it's equally true that if you have `mixed` appearing at the output end of a data flow, then Flow may stop caring what type is at the input end, and not notice even if *that* type has a type expression that should be an error.

Here, the `empty` is flowing simply to `number`. That will succeed, because `empty` can flow to anything. The test demonstrates that as a result, `{| +a: empty |}` can flow to `{| +a: number |}` (and that `IsSupertype` accurately captures that behavior.) On the other hand, `{| +a: mixed |}` can't flow to `{| +a: number |}`. That's because that would work only if `mixed` could flow to `number`, and it can't. That's even though if we needed the flow in the *other* direction, from `number` to `mixed`, it would succeed, because anything can flow to `mixed`.

This isn't real clear in the existing version of this file, I think. The discussion of `empty` at the top is a little overbroad, and there isn't a demonstration of how `mixed` can produce the same kind of short-circuit behavior when it's simply at the output end of the flow instead of the input end. I think I probably didn't understand that symmetry at the time I originally wrote that text; I spent some time digging into Flow this past December and January, and understand it a lot better now than I did before.

I think I'll merge this as is in this respect, but then try to make a further change to clarify.

Reply:
#5053